### ISSUE TITLE
Template-to-subject alignment on test dataset + segmentation issue

### ISSUE THREAD

**Issue by JoTisch** [m-dannecker/CINeMA#6]
Dear Maik,

Love your work and the clarity of your repository and documentation!

I trained the model for 1 epoch on the training dataset, then optimized it for 5–10 epochs while monitoring validation performance. The results look good. I was wondering:

- Do you plan to include a test script to generate template-to-subject alignments on an independent test dataset?
- Or did I miss an existing option for this?

In addition to my images, I provided segmentation masks (e.g., Background as BG, brain mask as totalBV). However, when I run training (see config snippet below), the model does not predict segmentation masks or mask the generated brain atlas. (DICE=1.0 and predicts label 1 for the complete volume)

- Do I need to enable the save_certainty_maps option to get this output?
- Are there any other config changes required to enable segmentation prediction?

Please let me know if I can provide any additional information.

Best, 
Johannes 

<details>
<summary>Atlas config here</summary>

```yaml
 root_dir: '/work/CINeMA' # if empty, use current working directory

# --------- Logging Parameters ---------
logging: False # if True, log to wandb
output_dir: ./CINeMA_results # directory to save the model, reconstruction images, atlases, etc.
wandb_entity: # wandb entity
project_name: CINeMA # wandb project name
config_data: dataset # name of the dataset to use (specified in config_data.yaml)

# --------- Device Parameters ---------
device: cuda
device_dataset: cuda
amp: True
seed: 42

# --------- General Options ---------
generate_cond_atlas: True # if True, generate the spatio-temporal (conditioned) atlas 
mask_reconstruction: True # mask inference reconstruction to remove noisy background
save_certainty_maps: False # if True, save the certainty maps (tissue probability maps)
compute_metrics: True # if True, compute the metrics for subject adaptation (requires GT references for each modality)
save_imgs: # if True, save the reconstruction images for the training/validation set
  train: False 
  val: False
save_model: True # if True, save the model to output_dir
load_model: 
  path: ""  
  epoch: 1 # which epoch to load

# --------- Data Augmentation Parameters ---------
# not used in the publication.
data_augmentation:
  activate: False
  # individual augmentations
  augment_deformation:
    p: 0.3
    num_control_points: 7
    max_displacement: 7.5
  augment_motion:
    p: 0.3
    degrees: 5
    translation: 5
    num_transforms: 4
  augment_noise: 
    p: 0.3
    mean: 1.5
    std: 3.0
  augment_biasfield:
    p: 0.3
    coeff: 1.0
  augment_gamma: 
    p: 0.3
    log_gamma: 0.4

# --------- Optimizer Parameters ---------
optimizer:
  re_init_latents: False # if True, re-initialize latent codes from gaussian distribution after each epoch
  lr_inr: 1.0e-4
  inr_weight_decay: 0.0
  lr_latent: 5.0e-4
  latent_weight_decay: 0.0
  lr_tf: 7.5e-3
  tf_weight_decay: 0.0
  tf_weight: 0.5
  loss_metric: 'l1'
  seg_weight: 1.0
  scheduler:
    type: 'cosine'
    eta_min: 2.5e-5


# --------- INR Decoder Parameters ---------
inr_decoder:
  tf_dim: 6 # 0 = off, 6 = rigid, 9 = rigid + scaling
  cnn_kernel_size: 3 # kernel size for the CNN for spatial modulation, if 0 no CNN is used
  latent_dim: [256, 3, 3, 3] # latent code dimension, [latent_dim, X, Y, Z]
  in_dim: 3 # input dimension of the coordinates, e.g. 3 for 3D images
  out_dim: [1, 2] #(number of modalities (not counting segmentation), number of segmentation channels)
  hidden_size: 1024 # hidden size of the MLP
  num_hidden_layers: 5 # number of hidden layers of the MLP
  modulated_layers: [1, 3, 5] # layers that are modulated by the latent code
  omega: [30, 30] # first and hidden omega of SIREN


# --------- Atlas Generation Parameters ---------
atlas_gen:
  spacing: [0.5, 0.5, 0.5] # spacing of the atlas in mm
  temporal_values: [20, 21, 22, 23, 24, 25, 26, 27, 28, 29] # steps in the temporal dimension (scan age_min, scan age_max)
  conditions: # conditions to use for the atlas generation (must be specified and activated in config_data.yaml)
    # scan_age:
    #   values: [1]
    #   normed_values: False
  # the condtions are combined with each temporal values --> exponential increase of combinations with each condition!!
    # birth_age: # condition name as stated in config_data.yaml
    #   values: [26, 31, 36, 41] # values to combine with the temporal values
    #   normed_values: False # if False, steps are specified in dataframe values, i.e., not normalized
    # lv_volume_normed:
    #   values: [-1.25, -0.75, 0, 0.75, 1.25]
    #   normed_values: True # if True, steps are specified in normalized values [-1, 1]
    
  # --- Mean Latent Regression ---
  gaussian_span: 1.0 # in weeks. +/- weeks of scan age to cover 95% of the weights, 
                     # i.e. 95% of the latent code of an atlas of age 30 weeks is made up 
                     # of latent codes from subjects with scan age 30 +/- 0.75 weeks, [29.25, 30.75]
  cond_scale: 0.05   # scaling factor for the condition vector 
                     # to adjust condition value range to latent code value range


# --------- Latent Space Analysis ---------
latent_anaylsis: # to predict scan_age from latentcode or explciit conditions during test-time
  activate: False
  predict_scan_age: 'NCA' #NCA, none for no prediction
  predict_birth_age: 'none' #NCA, none for no prediction, used for implicit birth-age prediction from latent code for ablation study.
  predict_ext_condition: 'none' #'MAE'  # predict a condition for test-time subjects, MAE for birth-age, binary crossentropy for sex, none for no prediction
  ba_regression:
    activate: False
    lr: 5.0e-4
    epochs: 1000


# --------- Training Parameters ---------
n_subjects: # number of subjects to use for training/validation
  train: 300
  val: 100
epochs: # number of epochs for training/validation
  train: 1
  val: 10 # the num epochs for latent code optimization, redone for each validation cycle)
validate_every: 1
batch_size: 250 # batch size for training/validation (number of subjects per batch)
n_samples: 25000 # number of coordinates in each "inner_batch" sampled across the subejct batch
num_workers: 32 # number of workers for dataloader

```
</details>

<details>

<summary>Dataset config</summary>

```yaml
dataset:
  dataset_name: dataset
  subject_ids: /work/CINeMA/configs/subject_ids.yaml
  tsv_file: /work/CINeMA/configs/ge_us_fetal.tsv
  modalities: ['MRI', 'Seg'] # modality names in the tsv file
  conditions:  # conditions with number of dimensions possibly used for atlas conditioning; conditions must be in constraints as lookup for normalization bounds
    scan_age: false
  constraints:  # sample only subjects that meet all constraints
    scan_age: 
      type: numeric
      min: 20
      max: 29
      distribution: 
        type: 'uniform_fillup'
        priority: 10 # higher priority means more important
  label_names: ['BG', 'totalBV']
  class_weights: [1.00, 1.0]
  world_bbox: [155, 155, 155] # size of the (roughly) smallest COMMON bounding box for all subjects in mm
  normalize_values: 'minmax' # 'minmax' or 'zscore'

```

</details>

**Comment by m-dannecker** [comment]
Hi Johannes,

Thanks for your interest in the work. Glad your initial results look good!

- You are absolutely right that independent inference script would be much more convenient for testing on new data. Unfortunately, so far I have not had time to integrate this. However, I'm currently not working on CINeMA. Perhaps in the upcoming months, I will add some additional functionality including the test-script.

- Looking at the code, I don't fully understand right now why this happens. I will try to reproduce this. 
Until then, you might try the following: Your problem might be related to the "halo" function in the pre-processing or to the "masking" int post-processing. To isolate it, I would suggest you to first set `mask_reconstruction: False # mask inference reconstruction to remove noisy background`. On top of that, you could try to add another label as 'BG' with the label_id=2. You would have to pre-process your mask to add a small halo (with label=2) around your brain (which is label=1). You can use the add_halo function for this: 
```
# add background halo to segmentation to allow masking of the brain in the postprsocessing step 
def add_background_halo(label_names, seg_nii, halo_width=1.5, background_label_str='BG'):
    seg = seg_nii.get_fdata()
    bg_label = label_names.index(background_label_str)
    mask_bg = (seg>0).astype(np.float32)
    mask_bg = (ndi.gaussian_filter(mask_bg, sigma=halo_width) > 0.001).astype(np.uint8) * bg_label
    mask_bg[seg > 0] = seg[seg > 0]
    return nib.Nifti1Image(mask_bg, seg_nii.affine)
```

Then you change your config to 
`  label_names: ['ClearLabel', 'totalBV', 'BG']
  class_weights: [0.0, 1.0, 1.0]
`
Sorry for this really inconvenient setup. But this way you won't have to wait until I have located the problem (hopefully sometime next week). 

Hope that helps!

Best,
Maik
